perception/autoware_compare_map_segmentation/src/voxel_distance_based_compare_map_filter/node.cpp

```diff
@@ -33,7 +33,7 @@ void VoxelDistanceBasedStaticMapLoader::onMapCallback(
   pcl::fromROSMsg<pcl::PointXYZ>(*map, map_pcl);
   const auto map_pcl_ptr = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>(map_pcl);
 
-  (*mutex_ptr_).lock();
+  std::lock_guard<std::mutex> lock(static_map_loader_mutex_);
```

Comment:
Why do you need a mutex here?
Other mutexes seem to protext `current_voxel_grid_dict_`

Reply:
this doesn't releated to `current_voxel_grid_dict_` protection in `dynamic loader mode`, 
This is to protect `voxel_grid_` while it is filtered in `static_map_loader` mode

Reply:
Then, please minimize the scope of mutex! 
It also applies to other parts.

Reply:
Thanks for your comment, I fixed at https://github.com/autowarefoundation/autoware.universe/pull/9097/commits/061359ea9a96060962b93e9e53af4c5f95337dd3.
PS: in this case, the mutex is better to cover `tree_` processing also.

Reply:
Let me confirm again:
- `dynamic_map_loader_mutex_` is a mutex for `current_voxel_grid_dict_`
- `static_map_loader_mutex_` is a mutex for `tree_ `

Is it correct?

Reply:
> dynamic_map_loader_mutex_ is a mutex for current_voxel_grid_dict_

Yes, it is for protecting voxelization (`voxel_grid_` update) in `dynamic_map_loader_ mode`

> `static_map_loader_mutex_` is a mutex for `tree_` 

Not only `tree_`, it is for protecting voxelization in `static_map_loader_` mode, in most option is `voxel_grid_`, and in some options `xx_distance_based_compare_map_filter`, it is also for `tree_` protection.
In `static_map_loader_` mode, the voxelization process is only occurred one time when Autoware starting, so I think it is acceptable to lock both 'voxel_grid_` and `tree_` together.

Reply:
Thank you for a detailed explanation. I have two questions:

1. Is it necessary to protect these two lines?
```
map_ptr_ = map_pcl_ptr;
*tf_map_input_frame_ = map_ptr_->header.frame_id;
```

2. I understand `voxel_grid_` and `tree_` are protected in initialization, but I wonder whether they are protected when being read afterwards.
For example,
```
  if (tree_ == NULL) {
    return false;
  }
```

Reply:
> but I wonder whether they are protected when being read afterwards.
For example,
  if (tree_ == NULL) {
    return false;
  }

Since we don't have any modification `tree_` afterwards, so the IMO the protection is unnecessary, how do think?

Reply:
> Is it necessary to protect these two lines?

Thanks, I fixed at https://github.com/autowarefoundation/autoware.universe/pull/9097/commits/01070934d56860d2c3636ac680f3f02cc413f489

Reply:
> Since we don't have any modification tree_ afterwards, so the IMO the protection is unnecessary, how do think?

Basically, this is correct.
However, to realize what you say, you need to implement a careful memory barrier or some mechanism like
- how to protect other threads to access `tree_` when a thread is initializing it
- how to guarantee that `tree_` is initialized when another thread tries to access `tree_`

If you do not understand what I say, please reference Memory Ordering.

Reply:
@veqcc Since `tree_`'s modification is only one time at begining, so I add its status checking before access instead of using mutex for more efficient as  at https://github.com/autowarefoundation/autoware.universe/pull/9097/commits/6824d74f5231550b59ede624a91e088b43111121

Reply:
@veqcc I undestood your point now, could you please review my modiffication at https://github.com/autowarefoundation/autoware.universe/pull/9097/commits/54e1da6da763d84df04326338d6879588d6fab18 and give me any suggestion you might have?

Reply:
LGTM for this point!! Thank you for a lot of your effort :+1: